lib/veri.ml

```diff
@@ -195,65 +204,32 @@ class ['a] t arch dis =
       | Bil.Mem _ | Bil.Bot -> SM.return ()
 
     method! eval_store ~mem ~addr data endian size =
-      super#eval_store ~mem ~addr data endian size >>= fun r ->
-      self#eval_exp addr >>= fun addr ->
-      self#eval_exp data >>= fun data ->
-      match value addr, value data with
-      | Bil.Imm addr, Bil.Imm data ->
-        SM.update (fun c -> c#discard_event (is_previous_store addr)) >>= fun () ->
-        let ev = create_mem_store addr data in
-        self#update_event ev >>= fun () -> SM.return r
-      | _ -> SM.return r
-
-    method private store_and_load ~mem ~addr mv endian size =
-      let data = Bil.int (Move.data mv) in
-      super#eval_store ~mem ~addr data endian size >>= fun r ->
-      match value r with
-      | Bil.Mem _ -> super#update mem_var r >>= fun () ->
-        super#eval_load ~mem ~addr endian size
-      | Bil.Imm _ | Bil.Bot -> SM.return r
-
-    (** [resolve_addr addr] - returns a result, bound with [addr].
-        Sequence of searches is the following:
-        1) among load events that occured at current step, in the same context,
-           with the same address;
-        2) among load events that occures at current step, in other context,
-           with the same address;
-        3) in current context. *)
-    method private resolve_addr ~mem ~addr endian size =
-      self#eval_exp addr >>= fun addr_res ->
-      match value addr_res with
-      | Bil.Bot | Bil.Mem _ -> SM.return addr_res
-      | Bil.Imm addr' ->
-        SM.get () >>= fun ctxt ->
-        match find_mem_load (self_events ctxt) (same_addr addr') with
-        | Some mv -> self#store_and_load ~mem ~addr mv endian size
-        | None ->
-          match find_mem_load (other_events ctxt) (same_addr addr') with
-          | None -> super#eval_load ~mem ~addr endian size
-          | Some mv -> self#store_and_load ~mem ~addr mv endian size
-
-    (** [eval_load ~mem ~addr endian size] - returns a result bound with [addr].
-        Search starts from self events, if it was write access to given
-        address at current step. And if it was, then result of write access
-        returned and no load event is emitted.
-        Otherwise searching continues as written above for [resolve_addr],
-        with emitting memory load event. *)
+      self#eval_exp addr >>= fun addr_r ->
+      self#eval_exp data >>= fun data_r ->
+      match value addr_r, value data_r with
+      | Bil.Imm got_addr, Bil.Imm got_data ->
+        SM.update (fun c -> c#discard_event (is_previous_store got_addr)) >>= fun () ->
```

Comment:
Let's continue the `discard` discussion (I can't reply to an outdated comment)

You were saying:

> `bil` user can easily write a code like 
> ```
> RAX := 0x42
> RAX := RAX + 1
> ```
> but what we need here is only one write event - last one

No, we need both events. We would like to have a stream of events that model all effects that occur during the evaluation of an expression.